### PR TITLE
chore(lint): Tweak and standardize ESLint configs

### LINKED ISSUE #12914: Clean up and standardize ESLint configs
_Originally posted by @clouserw in https://github.com/mozilla/fxa/issues/12820#issuecomment-1129271755_

> yeah, I think more work could be done here, but as one big patch it was feeling a bit unwieldy.  Some packages were linting js, but not other types.  There are several .eslintrc/ignore files scattered around the packages, some nested.  I think the right thing would be to go package by package and clean them up.

### PR DESCRIPTION
**Pull request by pdehaan**
## Because

- ESLint configs were a bit inconsistent and updating to follow best practices.

## This pull request

- Makes everything [incrementally] better. ❤️ 

## Issue that this pull request solves

Closes #12914 

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [ ] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).

## Screenshots (Optional)

Please attach the screenshots of the changes made in case of change in user interface.

## Other information (Optional)

```sh
yarn workspaces foreach -v run lint | grep -E " problems? "

➤ YN0000: [browserid-verifier]: ✖ 8 problems (0 errors, 8 warnings)
➤ YN0000: [fxa-admin-panel]: ✖ 2 problems (0 errors, 2 warnings)
➤ YN0000: [fxa-admin-server]: ✖ 6 problems (0 errors, 6 warnings)
➤ YN0000: [fxa-auth-client]: ✖ 4 problems (0 errors, 4 warnings)
➤ YN0000: [fxa-customs-server]: ✖ 5 problems (5 errors, 0 warnings)
➤ YN0000: [fxa-event-broker]: ✖ 21 problems (10 errors, 11 warnings)
➤ YN0000: [fxa-graphql-api]: ✖ 8 problems (1 error, 7 warnings)
➤ YN0000: [fxa-payments-server]: ✖ 1 problem (0 errors, 1 warning)
➤ YN0000: [fxa-react]: ✖ 1 problem (1 error, 0 warnings)
➤ YN0000: [fxa-settings]: ✖ 37 problems (13 errors, 24 warnings)
➤ YN0000: [fxa-support-panel]: ✖ 1 problem (1 error, 0 warnings)
```

```sh
yarn workspaces foreach --verbose run lint | grep "Process exited "

➤ YN0000: [123done]: Process exited (exit code 0), completed in 0s 661ms
➤ YN0000: [browserid-verifier]: Process exited (exit code 0), completed in 0s 691ms
➤ YN0000: [fortress]: Process exited (exit code 0), completed in 1s 383ms
➤ YN0000: [fxa-admin-panel]: Process exited (exit code 0), completed in 1s 601ms
➤ YN0000: [fxa-admin-server]: Process exited (exit code 0), completed in 1s 473ms
➤ YN0000: [fxa-auth-client]: Process exited (exit code 0), completed in 1s 395ms
➤ YN0000: [fxa-auth-server]: Process exited (exit code 0), completed in 2s 190ms
➤ YN0000: [fxa-content-server]: Process exited (exit code 0), completed in 4s 478ms
➤ YN0000: [fxa-customs-server]: Process exited (exit code 1), completed in 1s 56ms
➤ YN0000: [fxa-event-broker]: Process exited (exit code 1), completed in 1s 421ms
➤ YN0000: [fxa-geodb]: Process exited (exit code 0), completed in 0s 604ms
➤ YN0000: [fxa-graphql-api]: Process exited (exit code 1), completed in 1s 376ms
➤ YN0000: [fxa-payments-server]: Process exited (exit code 0), completed in 3s 603ms
➤ YN0000: [fxa-profile-server]: Process exited (exit code 0), completed in 0s 867ms
➤ YN0000: [fxa-react]: Process exited (exit code 1), completed in 0s 644ms
➤ YN0000: [fxa-settings]: Process exited (exit code 1), completed in 3s 222ms
➤ YN0000: [fxa-shared]: Process exited (exit code 0), completed in 0s 786ms
➤ YN0000: [fxa-support-panel]: Process exited (exit code 1), completed in 1s 126ms
```

### CODE REVIEW

**Comment by pdehaan** [comment]
TLDR:
- Renames everything from ".eslintrc" to ".eslintrc.json".
- Converts one file from YAML to JSON (for consistency).
- Adds a couple missing .eslintrc.json files.
- Fixes a couple of invalid ESLint configs.
- Converts from `0` and `2` to `warn` and `error`.

A few configs might need to change to .eslintrc.js in the future if we want to enable eslint-plugin-jest's [no-deprecated-functions](https://github.com/jest-community/eslint-plugin-jest/blob/main/docs/rules/no-deprecated-functions.md) rule and set a dynamic Jest version, per https://github.com/jest-community/eslint-plugin-jest#jest-version-setting.

File: packages/fxa-admin-server/.eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": [
+    "react-app",
```
Comment (LZoog) [inline packages/fxa-admin-server/.eslintrc.json:3]: I might be missing something but I don't think this should extend `react-app` since there's not any React in the admin-server (just the admin-panel). The auth-server extends `plugin:fxa/server`, maybe we use that?

File: packages/fxa-auth-server/scripts/.eslintrc.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (LZoog) [inline packages/fxa-auth-server/scripts/.eslintrc.json:1]: Small nit that's probably too much trouble to go and redo, to retain git history we might could've renamed these files with `git mv`. Seems like you either did that or git picked up on it with some other files but not all of them. If there's not a way to keep that history at this point without redoing most of them, then meh.

File: packages/fxa-event-broker/.eslintrc.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": [
+    "react-app",
```
Comment (LZoog) [inline packages/fxa-event-broker/.eslintrc.json:3]: This doesn't have any React in it either.

File: packages/fxa-support-panel/.eslintrc.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": [
+    "react-app",
```
Comment (LZoog) [inline packages/fxa-support-panel/.eslintrc.json:3]: Support panel doesn't have React either. Side note, we hope to [convert it to EJS](https://github.com/mozilla/fxa/issues/12998) for consistency 😎

File: yarn.lock [hunk 5]
```diff
@@ -3503,6 +3503,23 @@ __metadata:
   languageName: node
   linkType: hard
 
+"@eslint/eslintrc@npm:^1.3.0":
```
Comment (LZoog) [inline yarn.lock:3506]: What's this do for us that we couldn't do before? 👀

**Review by LZoog — COMMENTED** [review]
This is great, thanks @pdehaan. Mostly just have a couple questions around extending the React stuff. 🙂

File: packages/fxa-auth-client/.eslintrc.json [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": [
+    "react-app",
```
[the remark this answers is not in the document]
Comment (pdehaan) [reply packages/fxa-auth-client/.eslintrc.json:3]: Ah. I probably just saw that it had .ts files and figured react-app was the quickest way to set the correct parser+rules.

I'll see if I can find another package that lints TypeScript that doesn't have React since I'm not sure how well our "plugin:fxa/client" or "plugin:fxa/server" or "plugin:fxa/recommended" configs support TypeScript.

I was also hoping we could kill/refactor that old [eslint-plugin-fxa](https://github.com/mozilla/eslint-plugin-fxa/tree/master/lib/configs) and possibly [move it into this monorepo](https://github.com/mozilla/fxa/issues/13051) to easily share linting configs between projects without needing to do a npm publish (and then expand the configs to include "plugin:fxa/typescript", "plugin:fxa/react", "plugin:fxa/jest", "plugin:fxa/nestjs", "plugin:fxa/storybook", etc).

File: packages/fxa-admin-server/.eslintrc.json (hunk 1 above)
Comment (pdehaan) [reply packages/fxa-admin-server/.eslintrc.json:3]: OK, I think this will work for linting TypeScript w/ a similar config to react-app and react-app/jest (minus the React bits):

```json
{
  "env": {
    "es6": true,
    "node": true
  },
  "plugins": ["import"],
  "extends": ["eslint:recommended", "plugin:import/recommended"],
  "rules": {
    "import/no-unresolved": "off"
  },
  "root": true,
  "ignorePatterns": ["dist"],
  "overrides": [
    {
      "files": ["**/*.ts?(x)"],
      "parser": "@typescript-eslint/parser",
      "parserOptions": {
        "ecmaVersion": 2018,
        "sourceType": "module",
        "warnOnUnsupportedTypeScriptVersion": true
      },
      "plugins": ["@typescript-eslint"],
      "extends": ["plugin:@typescript-eslint/recommended"],
      "rules": {}
    },
    {
      "files": ["**/*.{spec,test}.*", "*.e2e-spec.*", "src/mocks.ts"],
      "env": {
        "jest/globals": true
      },
      "plugins": ["jest"],
      "extends": ["plugin:jest/recommended"],
      "rules": {}
    }
  ]
}
```

File: packages/fxa-auth-client/.eslintrc.json [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  "plugins": ["fxa"],
+  "rules": { "import/no-unresolved": "off" },
+  "root": true,
+  "ignorePatterns": ["dist"],
+  "overrides": [
+    {
+      "files": ["**/*.ts?(x)"],
+      "parser": "@typescript-eslint/parser",
+      "parserOptions": {
+        "ecmaVersion": 2018,
```
Comment (LZoog) [inline packages/fxa-auth-client/.eslintrc.json:13]: In some packages we have this set to `2020` and others `2018`. Should they be consistent or does setting this to `2020` break things?

**Review by LZoog — APPROVED** [review]
LGTM! Just had one question. Do you mind filing a follow up issue for possibly that `eslint-plugin-nestjs` that you found + to have custom shared configs for each project type as you also said in Slack? Would be awesome to do that after we get that style guide finalized and have a list of available fxa plugins/project types. 😎

File: yarn.lock (hunk 5 above)
Comment (pdehaan) [reply yarn.lock:3506]: I have no idea what https://github.com/eslint/eslintrc is, but definitely not something I would have installed; unless it came in as some sub-dependency of something else.

File: packages/fxa-auth-client/.eslintrc.json (hunk 7 above)
Comment (pdehaan) [reply packages/fxa-auth-client/.eslintrc.json:13]: Good catch, I'll try bumping them up to "2020" (to match fxa-auth-server) and see if linters gonna lint.

```sh
git grep -n '"ecmaVersion":' packages/**/.eslintrc* | cat

packages/fxa-auth-client/.eslintrc.json:13:        "ecmaVersion": 2018,
packages/fxa-auth-server/.eslintrc.json:8:    "ecmaVersion": "2020",
packages/fxa-event-broker/.eslintrc.json:12:        "ecmaVersion": 2018,
packages/fxa-graphql-api/.eslintrc.json:12:        "ecmaVersion": 2018,
packages/fxa-support-panel/.eslintrc.json:12:        "ecmaVersion": 2018,
```

Comment (pdehaan) [reply packages/fxa-auth-client/.eslintrc.json]: @LZoog Bumped everything to `2020` and re-ran the linters and everything passes.

File: packages/fxa-auth-server/.eslintrc.json [hunk 8]
```diff
@@ -5,7 +5,13 @@
   "extends": ["plugin:fxa/server"],
   "plugins": ["fxa"],
   "parserOptions": {
-    "ecmaVersion": "2020",
+    "ecmaVersion": 2020,
```
Comment (pdehaan) [inline packages/fxa-auth-server/.eslintrc.json:8]: Changed this from `"2020"` to `2020` since VSCode was saying `"2020"` wasn't a valid/recommended value.

File: packages/fxa-auth-server/.eslintrc.json [hunk 9]
```diff
@@ -5,7 +5,13 @@
...
     "sourceType": "module"
-  }
+  },
+  "ignorePatterns": [
+    "dist",
+    "fxa-*",
+    "vendor"
```
Comment (pdehaan) [inline packages/fxa-auth-server/.eslintrc.json:14]: Added "vendor" here since it was linting some vendor/ejs.js file, per https://github.com/mozilla/fxa/blob/397c5906816518658b821278513ce371bdad322a/packages/fxa-auth-server/scripts/install-ejs.sh.

File: packages/fxa-auth-server/package.json [hunk 10]
```diff
@@ -12,7 +12,7 @@
     "build": "yarn emails-scss && tsc --build && cp -R config public lib scripts dist/fxa-auth-server",
     "bump-template-versions": "node scripts/template-version-bump",
     "audit": "npm audit --json | audit-filter --nsp-config=.nsprc --audit=-",
-    "lint": "eslint . .storybook --ignore-pattern 'dist'",
+    "lint": "eslint . .storybook",
```
Comment (pdehaan) [inline packages/fxa-auth-server/package.json:15]: `'dist'` is now ignored via .eslintrc.json file's `ignorePatterns[]`.

**Comment by LZoog** [comment]
@pdehaan Sorry I hadn't merged this yet. I came back to do so and there's conflicts 🫤 Someone else can merge this for you tomorrow or Friday since I'm out, or I can on Tuesday when we're all back.

**Comment by LZoog** [comment]
@pdehaan Don't hate me, I just came back to check and there's more conflicts 🙈 Ping me on Slack when they're fixed this time so I can merge?

(Also, mind fixing up your commits into one? 😁)

**Comment by pdehaan** [comment]
Closing.
I'll try this again and do separate PRs for each package to avoid future merge conflicts.

**Comment by pdehaan** [comment]
# CHECKLIST
<kbd>git ls-files packages/*/package.json</kbd>

- [x] packages/123done/package.json #13625
- [x] packages/browserid-verifier/package.json #13626
- [ ] packages/db-migrations/package.json
- [x] packages/fortress/package.json #13628
- [ ] packages/functional-tests/package.json
- [ ] packages/fxa-admin-panel/package.json
- [ ] packages/fxa-admin-server/package.json
- [ ] packages/fxa-auth-client/package.json
- [x] packages/fxa-auth-server/package.json #13640
- [ ] packages/fxa-content-server/package.json
- [ ] packages/fxa-customs-server/package.json
- [ ] packages/fxa-dev-launcher/package.json
- [ ] packages/fxa-event-broker/package.json
- [ ] packages/fxa-geodb/package.json
- [ ] packages/fxa-graphql-api/package.json
- [ ] packages/fxa-payments-server/package.json
- [ ] packages/fxa-profile-server/package.json
- [ ] packages/fxa-react/package.json
- [ ] packages/fxa-settings/package.json
- [ ] packages/fxa-shared/package.json
- [ ] packages/fxa-support-panel/package.json
